### PR TITLE
Validate @Input(File(s)) ResolvedArtifactResult as unsupported

### PR DESCRIPTION
**Pull request by eskatos**
Previously, no validation was in place.

Build would fail on `@Input ResolvedArtifactResult` at task input fingerprinting stage.
But it would not fail with a task without outputs because then no fingerprinting was happening.

Build would fail on `@InputFile` because the value could not be resolved to a file.

Build would not fail with `@InputFiles`, resolving to an empty file collection and disabling task optimisations.

This PR adds validation of `@Input`, `@InputFile` and `@InputFiles` task properties to report usages of `ResolvedArtifactResult` as errors.

The validation happens both at plugin build time with `:validatePlugins` and at runtime.

See
* https://github.com/gradle/gradle/issues/20779
* https://github.com/gradle/gradle/issues/20785

### CODE REVIEW

**Comment by eskatos** [comment]
@bot-gradle test QFL

**Comment by eskatos** [comment]
@bot-gradle test this

**Comment by eskatos** [comment]
@bot-gradle test this

**Comment by eskatos** [comment]
@bot-gradle test this

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyManagementResultsAsInputsIntegrationTest.groovy [hunk 1]
```diff
@@ -148,8 +181,30 @@ class DependencyManagementResultsAsInputsIntegrationTest extends AbstractHttpDep
...
+        // Possible solutions
+        failureDescriptionContains("1. Extract artifact metadata and annotate with @Input.")
+        failureDescriptionContains("2. Extract artifact files and annotate with @InputFiles.")
+
+        // Documentation
+        failureDescriptionContains("Please refer to https://docs.gradle.org/${GradleVersion.current().version}/userguide/validation_problems.html#unsupported_value_type for more details about this problem.")
+
+        where:
+        annotation    | description
+        "@Input"      | "at-input"
```
Comment (wolfs) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyManagementResultsAsInputsIntegrationTest.groovy:205]: Is it bad to include an `@` in the unrolled test name? I'd just use the annotation directly.

**Review by wolfs — APPROVED** [review]
LGTM! I left one minor comment.

**Comment by eskatos** [comment]
@bot-gradle test and merge

**Pull request merged**